### ISSUE TITLE
Authentication Service using JWT allows unwanted resource access

### ISSUE THREAD

**Issue by luuk-ou** [cakephp/authentication#372]
Hey :) 

I have the following set up, using Cakephp 3.8 and the authentication plugin 1.x. Our project requires 3 types of authentication. Form Auth for web UI, Basic Auth for some controllers and JWT for other another set of controllers. So I decided to move the respective controller under their own prefixed routes in order to separate the authentication duties.

The following authentication services are configured:
1) Application.php -> contains AuthenticationService() using Authentication.Form for Web UI -> works fine
2) routes.php -> route::prefix('abc', ...  includes AuthenticationService() using Authentication.HttpBasic - > works fine too :)
3) routes.php -> route::prefix('def', .... includes AuthenticationService() using Authentication.Jwt -> does not work... seems to ignore any authentication configuration and allows access to the resource(s) under controller path '/def' ...

The code for both route setups look very similar, but I will show you the JWT one which doesn't work:
```
Router::prefix('def', function (RouteBuilder $routes) {
    $jwt_service = new AuthenticationService();
    $jwt_service->loadAuthenticator('Authentication.Jwt', [
        'returnPayload' => false
    ]);
    $jwt_service->loadIdentifier('Authentication.JwtSubject', [
        'resolver' => [
            'className' => 'Authentication.Orm',
            'userModel' => 'Users',
            'finder' => 'activeUserAccounts'
        ],
        'tokenField' => 'uuid',
        'dataField' => 'sub',
    ]);
    
    $routes->registerMiddleware('jwt_authentication', new AuthenticationMiddleware($jwt_service));
    $routes->applyMiddleware('jwt_authentication');
    
    // Connect API actions here.
    $routes->setExtensions(['json']);
    $routes->connect('/orders', ['controller' => 'Orders', 'action' => 'index', '_ext' => 'json']);
    $routes->fallbacks(DashedRoute::class);
});
```
So basically if you don't provide a JWT it allows access to the resources defined by prefix path. I would have expected a 401 or something similar to indicate unauthorized access.

Through debugging I noticed that the 'unauthorizedChallenge' of the TokenAuthenticator doesn't throw an error, like it does with HttpBasic, which means that the 'AuthenticationMiddleware' passes the request. Line 128 in this case returns, allowing the response to be exposed. 

Q? am I doing this wrong? Should I be using Authentication Service in a different manner? Anyway will fork the repo and do some further analysis. But I though I poss the question in the meantime in case I am on the wrong track. Any help appreciated :) Stay safe all.

**Comment by ndm2** [comment]
I suppose in your `Application::middleware()` method you are adding an instance of `AuthenticationMiddleware` (which uses session and form authentication) *after* `RoutingMiddleware`? (reversing the order could possibly fix it, but that wouldn't be the way to go!)

**Comment by luuk-ou** [comment]
Indeed my Application::middleware() is very much like suggested in the docs.... like so:

```
    public function middleware($middlewareQueue)
    {
        $combined_authentication = new AuthenticationMiddleware($this);
        
        $middlewareQueue
            // Catch any exceptions in the lower layers,
            // and make an error page/response
            ->add(new ErrorHandlerMiddleware(null, Configure::read('Error')))

            // Handle plugin/theme assets like CakePHP normally does.
            ->add(new AssetMiddleware([
                'cacheTime' => Configure::read('Asset.cacheTime')
            ]))

            // Add routing middleware.
            // If you have a large number of routes connected, turning on routes
            // caching in production could improve performance. For that when
            // creating the middleware instance specify the cache config name by
            // using it's second constructor argument:
            // `new RoutingMiddleware($this, '_cake_routes_')`
            ->add(new RoutingMiddleware($this))
            
            // Add authentication methods to middleware
            ->add($combined_authentication);

        return $middlewareQueue;
    }
```

I tried reversing the order just to see what happens. It actually breaks the form based authentication for the web UI and produces a 404. Which sorta makes sense as we are trying to authenticate against something that hasn't been established yet (the routes).

**Comment by ndm2** [comment]
And am I right to assume that you're either already session/form based authenticated when your testing the JWT auth (ie if you clear your cookies, accessing the `def` scope will trigger an error or redirect you), or you're not using the authentication component in your `def` scope (respectively its automatic identity check)?

**Comment by luuk-ou** [comment]
No... I am not authenticated by any means. Tested without cookies as well. Still gives me access to resources in def scope. Also not using Authetication component in def scope (yet).
________________________________

**Comment by ndm2** [comment]
Then there's your problem (or one of them).

JWT authentication relies on your application checking for an identity, and responding accordingly, just like it is with all the stateful authenticators. In most setups that would be an exception being thrown by the authentication component, unless you've configured it with `requireIdentity` disabled.

There is no standardized client based challenge for JWT tokens as there is for Basic credentials, hence the token authenticators do nothing in `StatelessInterface::unauthorizedChallenge()`.

If you implement such a check you'll probably still have a problem, unless you are restricting the authentication middleware set in your `Application` class to specific routes only, as otherwise you'd have two authentication middlewares running on every request.

**Comment by luuk-ou** [comment]
Thanks Oliver, that makes total sense. Having worked with the plugin admad/cakephp-jwt-auth in the past I was looking for similar behaviour I guess. From memory it throws an unauthenticatedException (not an unauthorizedChallenge) when no valid JWT is provided.

**Comment by ndm2** [comment]
You could move all the config in `Application::getAuthenticationService()` and apply it based on the routing parameters, like:

```php
public function getAuthenticationService(
    ServerRequestInterface $request,
    ResponseInterface $response
) {
    $params = $request->getAttribute('params');
    $prefix = isset($params['prefix']) ? $params['prefix'] : null;

    $service = new AuthenticationService();
    switch ($prefix) {
        case 'abc':
            // set Abc prefix specific config
            break;

        case 'def':
            // set Def prefix specific config
            break;

        default:
            // set default config
            break;
    }
    

    return $service;
}
```

Another option would be to move everything into routing scopes, like:

```php
$routes->scope('/', function (RouteBuilder $routes) {
    $service = new AuthenticationService();
    // set default config
    
    $routes->registerMiddleware('defaultAuth', new AuthenticationMiddleware($service));
    $routes->applyMiddleware('defaultAuth');

    // ...
});

$routes->prefix('Abc', function (RouteBuilder $routes) {
    $service = new AuthenticationService();
    // set Abc prefix specific config

    $routes->registerMiddleware('abcSpecificAuth', new AuthenticationMiddleware($service));
    $routes->applyMiddleware('abcSpecificAuth');

    // ...
});

$routes->prefix('Def', function (RouteBuilder $routes) {
    $service = new AuthenticationService();
    // set Def prefix specific config

    $routes->registerMiddleware('defSpecificAuth', new AuthenticationMiddleware($service));
    $routes->applyMiddleware('defSpecificAuth');

    // ...
});

**Comment by luuk-ou** [comment]
That's awesome Oliver. I took the approach of using the `Application::getAuthenticationService()` as you suggested. That keeps the `routes.php` nice and tidy. Additionally I configured the `AuthenticationService()` to not perform the redirect, i.e. `'unauthenticatedRedirect' => false`, which is the behaviour I want. Thanks again for your guidance. Much appreciated.
